### PR TITLE
KAFKA-15326: [7/N] Processing thread non-busy waiting

### PR DESCRIPTION
**Pull request by lucasbru**
Avoid busy waiting for processable tasks. We need to be a bit careful here to not have the task executors to sleep when work is available. We have to make sure to signal on the condition variable any time a task becomes "processable". Here are some situations where a task becomes processable:

- Task is unassigned from another `TaskExecutor`.
- Task state is changed (should only happen inside when a task is locked inside the polling phase).
- When tasks are unlocked.
- When tasks are added.
- New records available.
- A task is resumed.

So in summary, we

- We should probably lock tasks when they are paused and unlock them when they are resumed. We should also wake the task executors after every polling phase. This belongs to the StreamThread integration work (separate PR). We add `DefaultTaskManager.signalProcessableTasks` for this.
- We need to awake the task executors in `DefaultTaskManager.unassignTask`, `DefaultTaskManager.unlockTasks` and `DefaultTaskManager.add`.

### CODE REVIEW

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskManager.java [hunk 1]
```diff
@@ -118,6 +120,41 @@ public StreamTask assignNextTask(final TaskExecutor executor) {
...
 
+    @Override
+    public void awaitProcessableTasks() throws InterruptedException {
+        final boolean interrupted = returnWithTasksLocked(() -> {
+            for (final Task task : tasks.activeTasks()) {
+                if (!assignedTasks.containsKey(task.id()) &&
+                    !lockedTasks.contains(task.id()) &&
+                    canProgress((StreamTask) task, time.milliseconds())
+                ) {
+                    log.info("Await unblocked: returning early from await since a processable task {} was found", task.id());
```
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskManager.java]: I think info might be too high of a log level. This should be called frequently right?

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/TaskManager.java [hunk 2]
```diff
@@ -121,4 +119,14 @@ public interface TaskManager {
      */
     Map<TaskId, StreamsException> drainUncaughtExceptions();
 
+    /**
+     * Signals that at least one task has become processable, e.g. because it was resumed or new records may be available.
```
Comment (wcarlson5) [inline streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/TaskManager.java:123]: I assume this method with be used in later commits?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskManagerTest.java [hunk 3]
```diff
@@ -94,6 +103,114 @@ public void shouldAssignTaskThatCanBeProcessed() {
...
+        }
+    }
+
+    @Test
+    public void shouldBlockOnAwait() throws InterruptedException {
+        final AwaitingRunnable awaitingRunnable = new AwaitingRunnable();
+        final Thread awaitingThread = new Thread(awaitingRunnable);
+        awaitingThread.start();
+
+        assertFalse(awaitingRunnable.awaitDone.await(100, TimeUnit.MILLISECONDS));
```
Comment (wcarlson5) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskManagerTest.java:132]: Do you want to use the `VERIFICATION_TIMEOUT` here?

**Review by wcarlson5 — COMMENTED** [review]
Looking good, just a few comments/questions

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskManagerTest.java (hunk 3 above)
Comment (lucasbru) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskManagerTest.java:132]: Thanks, yes.

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/TaskManager.java (hunk 2 above)
Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/TaskManager.java:123]: Yes!

File: streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskManager.java (hunk 1 above)
Comment (lucasbru) [reply streams/src/main/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskManager.java]: Yes, you are right. I changed it in a later PR, but I pulled it back into this PR.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutorTest.java [hunk 4]
```diff
@@ -85,6 +85,16 @@ public void shouldShutdownTaskExecutor() {
...
+    @Test
+    public void shouldAwaitProcessableTasksIfNoneAssignable() throws InterruptedException {
+        assertNull(taskExecutor.currentTask(), "Have task assigned before startup");
+        when(task.isProcessable(anyLong())).thenReturn(false);
+
+        taskExecutor.start();
+
+        verify(taskManager, timeout(VERIFICATION_TIMEOUT).atLeastOnce()).awaitProcessableTasks();
+    }
+
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutorTest.java]: This test is a bit hard to follow. I think the test would be easier to understand if you specify the simpler stub `when(taskManager.assignNextTask(taskExecutor)).thenReturn(null);` here in the test instead of using the stub specified in `setUp()`.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskManagerTest.java [hunk 5]
```diff
@@ -94,6 +103,114 @@ public void shouldAssignTaskThatCanBeProcessed() {
...
+        }
+
+        public void shutdown() {
+            shutdownRequested.set(true);
+            taskManager.signalProcessableTasks();
+        }
+    }
+
+    @Test
+    public void shouldBlockOnAwait() throws InterruptedException {
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskManagerTest.java:127]: Waiting for the whole verification timeout for a successful test result seems long to me.  
Does `shouldReturnFromAwaitOnInterruption()` not imply that the thread is waiting? Do we miss anything if we remove `shouldBlockOnAwait()`?

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskManagerTest.java [hunk 6]
```diff
@@ -73,6 +80,8 @@ public void setUp() {
         when(task.id()).thenReturn(new TaskId(0, 0, "A"));
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskManagerTest.java]: nit: 
You could reuse `org.apache.kafka.test.StreamsTestUtils.TaskBuilder.statelessTask` here instead of creating your own stubs.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskManagerTest.java [hunk 7]
```diff
@@ -94,6 +103,114 @@ public void shouldAssignTaskThatCanBeProcessed() {
...
+        awaitingRunnable.shutdown();
+    }
+
+    @Test
+    public void shouldReturnFromAwaitOnUnassignment() throws InterruptedException {
+        taskManager.add(Collections.singleton(task));
+        when(taskExecutionMetadata.canProcessTask(eq(task), anyLong())).thenReturn(true);
+
+        final StreamTask t = taskManager.assignNextTask(taskExecutor);
+        assertNotNull(t);
```
Comment (cadonna) [inline streams/src/test/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskManagerTest.java]: nit:
```suggestion
        final StreamTask task = taskManager.assignNextTask(taskExecutor);
        assertNotNull(task);
```

**Review by cadonna — COMMENTED** [review]
@lucasbru Thanks for the PR!

I really like the unit tests for `awaitProcessableTasks()`.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskManagerTest.java (hunk 5 above)
Comment (lucasbru) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskManagerTest.java:127]: I think it would be nice to keep it. But yeah, now I realize why I set this to 100. How about keeping it with 100? `shouldReturnFromAwaitOnInterruption` could pass if we always return immediately.

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskManagerTest.java (hunk 6 above)
Comment (lucasbru) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskManagerTest.java]: Done

File: streams/src/test/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutorTest.java (hunk 4 above)
Comment (lucasbru) [reply streams/src/test/java/org/apache/kafka/streams/processor/internals/tasks/DefaultTaskExecutorTest.java]: Done

**Review by cadonna — APPROVED** [review]
LGTM!

Thanks for the PR, @lucasbru !

**Comment by cadonna** [comment]
Build failures are unrelated.
